File: src/main/java/com/hiberus/anaya/redmineeditor/utils/Desktop.java
package com.hiberus.anaya.redmineeditor.utils;

import java.io.IOException;
import java.net.URI;
import java.net.URISyntaxException;

public class Desktop {

    /**
     * Open a page in the browser
     *
     * @param url page to open
     * @return true if it was opened, false on error
     */
    public static boolean openInBrowser(String url) {
        if (java.awt.Desktop.isDesktopSupported() && java.awt.Desktop.getDesktop().isSupported(java.awt.Desktop.Action.BROWSE)) {
            try {
                java.awt.Desktop.getDesktop().browse(new URI(url));
                return true;
            } catch (IOException | URISyntaxException e) {
                e.printStackTrace();
                return false;
            }
        } else {
            return false;
        }
    }
}

File: src/main/java/com/hiberus/anaya/redmineeditor/utils/Desktop.kt
package com.hiberus.anaya.redmineeditor.utils

import java.awt.Desktop.*
import java.io.File
import java.net.URI

/**
 * Open this URI in the browser
 * @return true if it was opened, false on error
 */
fun URI.openInBrowser() =
    if (!isDesktopSupported() || !getDesktop().isSupported(Action.BROWSE)) {
        // not supported
        false
    } else runCatching {
        // browse
        daemonThread { getDesktop().browse(this) }
    }.onFailure { it.printStackTraceFix() }.isSuccess

/**
 * Opens this file in an external app/editor
 * @return true if it was opened, false on error
 */
fun File.openInApp() =
    if (!exists()) {
        // file doesn't exist
        false
    } else if (!isDesktopSupported() || !getDesktop().isSupported(Action.OPEN)) {
        // not supported
        false
    } else runCatching {
        // open
        daemonThread { getDesktop().open(this) }
    }.onFailure { it.printStackTraceFix() }.isSuccess
